File: AdventOfCode/2019/22/22b.py
def modinv(a, m):
    # Works because m is prime
    return pow(a, m - 2, m)

lines = list(reversed(open("input").read().splitlines()))
CARDS = 119315717514047
TIMES = 101741582076661

def next(n):

    for line in lines:
        assert(0 <= n < CARDS)
        tokens = line.split()
        if tokens[0] == 'cut':
            k = int(tokens[1])
            n = (n + k + CARDS) % CARDS
    
        elif tokens[1] == 'into':
            n = CARDS - 1 - n

        elif tokens[1] == 'with':
            k = int(tokens[3])
            n = (modinv(k, CARDS) * n) % CARDS

        else:
            assert(False)
        assert(0 <= n < CARDS)
    return n

p0 = 2020
p1 = next(p0)
p2 = next(p1)
incr = ((p1 - p2) * modinv(p0 - p1 + CARDS, CARDS)) % CARDS
offs = (p1-incr*p0) % CARDS
print(incr, offs)

assert((incr*p0 + offs) % CARDS == next(p0))
assert((incr*p1 + offs) % CARDS == next(p1))

res = pow(incr, TIMES, CARDS)*p0 + (pow(incr, TIMES, CARDS) - 1) * modinv(incr - 1, CARDS) * offs
print(res % CARDS)

File: AdventOfCode/2019/22/22b.cc
#include <bits/stdc++.h>
using namespace std;

#define FOR(a, b, c) for (auto(a) = (b); (a) < (c); ++(a))

typedef long long ll;

char operation[100];

#define INC 0
#define CUT 1
#define STACK 2

#define ITER_AMOUNT 101741582076661LL
#define CARDS 119315717514047LL

struct op {
    int id;
    ll amount;
};

ll modpow(ll b, ll e, ll m) {
    if(e == 0) 
        return 1;

    if(e % 2 == 0)
        return modpow((b*b) % m, e / 2, m);
    
    return (b * modpow(b, e - 1, m)) % m;
}

ll modinv(ll a, ll m) {
    //Because m is prime
    return modpow(a, m - 2, m);
}

vector<op> ops;

ll next(ll pos) {
    for(op curr : ops) {
        switch(curr.id) {
            case CUT:
                pos = (pos + curr.amount + CARDS) % CARDS;
                break;
            case STACK:
                pos = CARDS - 1 - pos;
                break;
            case INC:
                //FIXME: Overflow: 
                pos = (modinv(curr.amount, CARDS) * pos) % CARDS;
                assert(pos >= 0);
                break;
            default:
                assert(false);
        }
    }

    return pos;
}
int main() {

    while(scanf(" %100[^\n]", operation) != EOF) {
        if(operation[0] == 'd') {
            if(operation[5] == 'w') {
                int increment;
                sscanf(operation, "deal with increment %d", &increment);
                ops.push_back({INC, increment});
            } else {
                assert(operation[5] == 'i');
                ops.push_back({STACK, -100000});
            }
        } else {
            assert(operation[0] == 'c');
            int cutAmount;
            sscanf(operation, "cut %d", &cutAmount);
            ops.push_back({CUT, cutAmount});
        }
    }
    reverse(ops.begin(), ops.end());

    // p = a * p_old + b
    ll pos = 2020;
    
    ll X = next(pos);
    ll Y = next(X);
    ll Z = next(Y);

    printf("%lld, %lld, %lld", X, Y, Z);

    return 0;
}